src/SmartEntity/FlexEntity.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/*
+ * This file is part of the Bushido\Foundation package.
+ *
+ * (c) Wojciech Nowicki <wnowicki@me.com>
+ *
+ * For the full copyright and license information, please view the LICENSE.md
+ * file that was distributed with this source code.
+ */
+
+namespace Bushido\Foundation\SmartEntity;
+
+use Bushido\Foundation\Contracts\Makeable;
+use Bushido\Foundation\Exceptions\InvalidArgumentException;
+use Bushido\Foundation\Helpers\ArrayableHelper;
+use Bushido\Foundation\Helpers\ChangeCase;
+
+class FlexEntity implements Entity
+{
+    private $data = [];
+
+    /**
+     * @param array $components
+     * @return static
+     */
+    public static function make(array $components): Makeable
+    {
+        $entity = new static();
+
+        foreach ($components as $k => $v) {
+            $entity->{'set' . ChangeCase::snakeToCamel($k)}($v);
+        }
+        return $entity;
+    }
+
+    public function toArray(): array
+    {
+        return ArrayableHelper::processArray($this->data);
+    }
+
+    public function toJson(int $options = 0): string
+    {
+        return (string) json_encode($this->toArray(), $options);
+    }
+
+    public function __toString(): string
+    {
+        return $this->toJson();
+    }
+
+    public function __call(string $name, array $arguments)
+    {
+        $action = substr($name, 0, 3);
+        $property = ChangeCase::camelToSnake(substr($name, 3));
+        switch ($action) {
+            case 'set':
+                return $this->set($property, $arguments);
+            case 'get':
+                return $this->get($property);
+            case 'add':
+                return $this->add($property, $arguments);
+        }
+        throw new \RuntimeException('Call to undefined method '.__CLASS__.'::'.$name.'()');
+    }
+
+    /**
+     * @param string $property
+     * @param array $arguments
+     * @return $this
+     */
+    private function set(string $property, array $arguments)
+    {
+        $this->checkArguments($arguments, $property);
+        $this->data[$property] = $arguments[0];
+
+        return $this;
+    }
+
+    private function get(string $property)
+    {
+        if (array_key_exists($property, $this->data)) {
+            return $this->data[$property];
+        }
+
+        return null;
+    }
+
+    private function add(string $name, array $arguments)
+    {
+        if (!isset($this->data[$name]) || is_array($this->data[$name])) {
+            $this->data[$name][] = $arguments[0];
+
+            return $this;
+        }
+
+        throw new InvalidArgumentException('Expected property [' . $name . '] to be array type');
```

Comment:
I think you've missed interpreted what is expected there, `$arguments` are always `array` as this is the use of a magic function. So hint is right and has nothing to do with the value itself, value is `$arguments[0]`.

I'm referring there to the type of property. I can change but message but I made less sense to me. In the `FlexEntity` we are only checking if we are not trying to add a new cell to non-array property, it can always be any value.

Reply:
For the first part, I missed the magic function and the fact it'll always be an array. The message is still not clear enough I think, when reading it properly it makes perfect sense but anyone passing an array to the `add()` function might be confused, consider the following:

```php
$entity = FlexEntity::make(['foo' => '']);

$entity->addFoo('bar');
// Bushido\Foundation\Exceptions\InvalidArgumentException: Expected property [foo] to be array type

// Okay, so I can't add "bar", maybe I can add ['bar']
$entity->addFoo(['bar']);
// Bushido\Foundation\Exceptions\InvalidArgumentException: Expected property [foo] to be array type
```

```php
$entity = FlexEntity::make(['foo' => '']);

$entity->addFoo('bar');
// Bushido\Foundation\Exceptions\InvalidArgumentException: The property [foo] needs to be an array to be able to add any values
```

Reply:
I'm not sure about the exact wording, but I'm on board with improving on it.
Maybe:
```
// Bushido\Foundation\Exceptions\InvalidArgumentException: Cannot add to [foo] property as it is not an array.
```
?